File: examples/governance/compliance_domain/lib/compliance_domain/governance_policy/commands/create_policy.rb
module ComplianceDomain
  class GovernancePolicy
    module Commands
      class CreatePolicy
        include Hecks::Command
        emits "CreatedPolicy"

        attr_reader :name
        attr_reader :description
        attr_reader :category
        attr_reader :framework_id

        def initialize(
          name: nil,
          description: nil,
          category: nil,
          framework_id: nil
        )
          @name = name
          @description = description
          @category = category
          @framework_id = framework_id
        end

        def call
          GovernancePolicy.new(
            name: name,
            description: description,
            category: category,
            framework_id: framework_id,
            status: "draft"
          )
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/create_policy.go
package domain

import (
	"time"
)

type CreatePolicy struct {
	Name string `json:"name"`
	Description string `json:"description"`
	Category string `json:"category"`
	FrameworkId string `json:"framework_id"`
}

func (c CreatePolicy) CommandName() string { return "CreatePolicy" }

func (c CreatePolicy) Execute(repo GovernancePolicyRepository) (*GovernancePolicy, *CreatedPolicy, error) {
	agg := NewGovernancePolicy(c.Name, c.Description, c.Category, c.FrameworkId, time.Time{}, time.Time{}, nil, "")
	agg.Status = "draft"
	if err := agg.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(agg); err != nil {
		return nil, nil, err
	}
	event := CreatedPolicy{
		AggregateID: agg.ID,
		Name: c.Name,
		Description: c.Description,
		Category: c.Category,
		FrameworkId: c.FrameworkId,
		OccurredAt: time.Now(),
	}
	return agg, &event, nil
}
